### DISCUSSION TITLE
Easy massive speed-up for multiple functions in algorithms.link_prediction

### DISCUSSION THREAD

**Discussion by marbinner** [networkx/networkx#7243] [Ideas]
Some functions need to know the neighborhood overlap of 2 nodes.

For example
https://networkx.org/documentation/latest/_modules/networkx/algorithms/link_prediction.html#common_neighbor_centrality
or
https://networkx.org/documentation/latest/_modules/networkx/algorithms/link_prediction.html#jaccard_coefficient

In these examples it's implemented using 
`len(list(nx.common_neighbors(G, u, v)))`
or
`sum(1 for _ in nx.common_neighbors(G, u, v))`

This is very slow however and makes the respective code basically unusable for the graphs I'm dealing with. One small change could easily fix this. Instead of getting the number of elements in a generator we could use set intersection:

```python
def overlap1(G, u, v):
    return len(set(nx.neighbors(G,u)) & set(nx.neighbors(G,v)))
```

I'd be happy to hear thoughts on this.

**Comment by MridulS** [comment]
Thanks for this! Maybe we can just speed up `nx.common_neighbors` directly! Currently it returns a generator with a for loop. I wonder if it makes sense to just to use the `G._adj` dictionary directly for set operations!

Some quick benchmarks for a graph with 10,000 nodes.
``` python


In [21]: %%time
    ...: _ = set(nx.common_neighbors(G, 1, 2))
    ...:
    ...:
CPU times: user 2.74 ms, sys: 3 µs, total: 2.74 ms
Wall time: 2.78 ms

In [22]: %%time
    ...: _ = set(nx.neighbors(G, 1)) & set(nx.neighbors(G, 2))
    ...:
    ...:
CPU times: user 515 µs, sys: 55 µs, total: 570 µs
Wall time: 575 µs

In [23]: %%time
    ...: _ = G._adj[1].keys() & G._adj[2].keys()
    ...:
    ...:
CPU times: user 202 µs, sys: 33 µs, total: 235 µs
Wall time: 240 µs

```

**Reply by marbinner** [reply]
While on the topic there seems to be another function that might get a similar speedup:
`nx.non_neighbors` could use set difference:  `G.nodes() - nbors`

**Reply by MridulS** [reply]
Yes! There is an open PR https://github.com/networkx/networkx/pull/7244 which implements these functions using the native `G._adj` dictionary which should be quicker than using the methods like `G.nodes()` :D

**Comment by rossbar** [comment]
#7244 is merged so this has hopefully been fully addressed - if there are further ideas for improvements, go ahead and open a new issue and link back to this one!
